File: src/main/java/br/com/beleza/service/ProductServiceImpl.java
package br.com.beleza.service;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import br.com.beleza.model.Product;
import br.com.beleza.repository.ProductRepository;

@Service
public class ProductServiceImpl implements ProductService {

	@Autowired
	private ProductRepository repository;
	
	@Override
	public void save(Product product) {
		repository.save(product);
	}

	@Override
	public void update(Product product) {
		repository.update(product);
	}

	@Override
	public void delete(Product product) {
		repository.delete(product);
	}

	@Override
	public Product find(Integer sku) {
		return repository.find(sku);
	}

}

File: src/main/kotlin/br/com/blz/testkotlin/service/ProductServiceImpl.kt
package br.com.blz.testkotlin.service

import br.com.blz.testkotlin.entity.ProductEntity
import br.com.blz.testkotlin.repository.ProductRepository
import org.springframework.stereotype.Service

@Service
class ProductServiceImpl(val productRepository: ProductRepository) : ProductService {

  override fun saveProduct(product: ProductEntity): ProductEntity? {
    productRepository.save(product)
    return getProductBySku(product.sku)
  }

  override fun editProduct(product: ProductEntity): ProductEntity? {
    productRepository.update(product)
    return getProductBySku(product.sku)
  }

  override fun getProductBySku(sku: Long): ProductEntity? {
    var product = productRepository.findBySku(sku)
    if (product != null) {
      val sum =
        product.inventory.warehouses.map { warehouse -> warehouse.quantity }.reduceOrNull { sum, quantity -> sum + quantity }
      product.inventory.quantity = sum ?: 0
      product.isMarketable = product.inventory.quantity > 0
    }
    return product
  }

  override fun deleteProductBySku(sku: Long): Boolean {
    return productRepository.delete(sku)
  }

  override fun getAllProducts(): MutableList<ProductEntity>? {
    return productRepository.getAll()
  }
}
